### PR TITLE
Add initial support in the sidecar for feature flags

### LINKED ISSUE #75: Add initial support for LaunchDarkly feature flags
### Discussed in https://github.com/confluentinc/vscode/discussions/353

This issue covers the initial support for using LaunchDarkly feature flags to the sidecar, and https://github.com/confluentinc/vscode/issues/355 covers adding initial support to the Confluent for VS Code extension itself.

Feature flags will be added and use with other issues.

### PR DESCRIPTION
**Pull request by rhauch**
## Summary of Changes

Fixes #75

This adds the infrastructure for using LaunchDarkly feature flags and exposing a new REST API to allow the VS Code extension to get the current evaluated value for a specified feature flag.

## Any additional details or context that should be provided?

LaunchDarkly does not yet have a Java client SDK that uses a non-secret client ID. (There is a Java server SDK, but this is intended as a multitenant services and requires an API key and secret that should not be shared.)

This framework uses a simple HTTP client to poll the HTTP API to get the currently evaluated flags given a context. The flags are evaluated on a schedule (e.g., every 60 seconds) and cached. If the sidecar is unable to evaluate the flags the first time (e.g., LaunchDarkly is not available when the sidecar starts), a set of default flags is used. However, as soon as flags can be evaluated the results are cached. Subsequent failures to refresh the flags (e.g., LaunchDarkly goes down after the sidecar process was started) will simply keep the cached flags (there is no TTL).

This framework also allows code within the sidecar to easily check feature flags. Simply `@Inject` a `FeatureFlags` reference, and call any of the methods to get feature flag values given a `FlagId` enum literal. There are default values defined for all `FlagId`s (verified by unit tests), which means the methods will always return a value for those feature flags.

To add a feature flag:
1. Define the flag in LaunchDarkly
2. Add a literal to the `FlagId` enum
3. Add a default value to the `src/main/resources/feature-flag-defaults.json` file
4. Use the flag in your code by calling a `FeatureFlags.evaluateAs<ValueType>()` method 
5. Profit!

Because the sidecar version is needed in the device context, this also refactors how that information is kept (moving it out of the `VersionResource` and into a new `SidecarInfo` application bean with several other attributes. This bean and the various attributes can be injected into other beans, including the new `FeatureFlags`.

Currently the OS information is read from the System properties, and the VS Code version and extension version information is read from either System properties (lowercase dot-delimited), such as
```
ide-sidecar-*-runner -Dvscode.version=1.93.2 -Dvscode.extension.version=v0.17.1 …
```
or if those are absent from environment variables (capital case underscore-delimited) used to start the sidecar process, such as:
```
VSCODE_VERSION=1.93.1 VSCODE_EXTENSION_VERSION=v0.17.0 ide-sidecar-*-runner …
```
These are currently optional, but we do want to change the extension to start supplying them soon, as the VS Code information is passed in the `device` contexts passed to LaunchDarkly during flag evaluation.

## Pull request checklist

Please check if your PR fulfills the following (if applicable):

- Tests:
    - [x] Added new
    - [x] Updated existing
    - [ ] Deleted existing
- [x] Have you validated this change locally against a running instance of the Quarkus dev server?
    ```shell
    make quarkus-dev
    ```
- [x] Have you validated this change against a locally running native executable?
    ```shell
    make mvn-package-native && ./target/ide-sidecar-*-runner
    ```

### CODE REVIEW

**Comment by rhauch** [comment]
Users should see very little difference in the sidecar logs. They may see errors if there are problems parsing the LD results, but otherwise the `FeatureFlags` class and its supporting classes are pretty silent during operation. We don't want log problems connecting to LaunchDarkly, since a user might not have internet connectivity. 

There are quite a few debug messages that we can see during development, if we enable debug logs.

Anyway, here's an example startup:
```
$ ./target/ide-sidecar-0.26.0-runner -Dvscode.version=v1.93.2 -Dvscode.extension.version=v0.17.1
```
with resulting log:
```
2024-10-09 17:09:47,342 INFO  [io.con.ide.res.app.LifecycleBean] (main) Sidecar starting...
2024-10-09 17:09:47,400 INFO  [io.con.ide.res.app.BeanProducers] (main) Extracted templates to /var/folders/yf/kdn74cm97_z5_jdbx_cygb_m0000gq/T/12022813187248413278
2024-10-09 17:09:47,405 INFO  [io.con.ide.res.app.SidecarInfo] (main) OS: Mac OS X 13.7 (MacOS); VS Code 1.93.2, extension version 0.17.1
2024-10-09 17:09:47,405 INFO  [io.con.ide.res.fea.FeatureFlags] (main) Checking for current feature flags in projects [IDE, Confluent Cloud], using 5 defaults, 0 overrides
2024-10-09 17:09:47,409 INFO  [io.con.ide.res.fea.FeatureFlags] (main) Next evaluation of feature flags will be after 5:10:47 PM (in 60s)
2024-10-09 17:09:47,543 INFO  [io.quarkus] (main) ide-sidecar 0.26.0 native (powered by Quarkus 3.13.3) started in 0.230s. Listening on: http://127.0.0.1:26636
2024-10-09 17:09:47,543 INFO  [io.quarkus] (main) Profile prod activated. 
2024-10-09 17:09:47,543 INFO  [io.quarkus] (main) Installed features: [cdi, config-yaml, hibernate-validator, qute, qute-web, reactive-routes, rest, rest-jackson, scheduler, smallrye-context-propagation, smallrye-graphql, smallrye-health, smallrye-openapi, vertx]
2024-10-09 17:09:48,002 INFO  [io.con.ide.res.app.KnownWorkspacesBean] (executor-thread-2) Grim reaper: Will check if the sidecar should remain alive periodically.
```
There are just a few new log messages. The third line in the output (repeated below) is new:
```
2024-10-09 17:09:47,405 INFO  [io.con.ide.res.app.SidecarInfo] (main) OS: Mac OS X 13.7 (MacOS); VS Code 1.93.2, extension version 0.17.1
```
Note that if the startup command excludes the `-Dvscode.version=v1.93.2 -Dvscode.extension.version=v0.17.1`:
```
$ ./target/ide-sidecar-0.26.0-runner
```
then this log message becomes:
```
2024-10-09 17:11:26,240 INFO  [io.con.ide.res.app.SidecarInfo] (main) OS: Mac OS X 13.7 (MacOS); VS Code unknown, extension version unknown
```
and the VS Code version and extension version information is not included in the LD contexts during feature flag evaluation.

The other new log message is this:
```
2024-10-09 17:09:47,409 INFO  [io.con.ide.res.fea.FeatureFlags] (main) Next evaluation of feature flags will be after 5:10:47 PM (in 60s)
```
which will occur again roughly every 60 seconds, or whenever the CCloud connection authenticates or disconnects (since this changes the LD context used for feature flag evaluations).

File: src/main/java/io/confluent/idesidecar/restapi/featureflags/FeatureFlags.java
Comment (flippingbits) [inline src/main/java/io/confluent/idesidecar/restapi/featureflags/FeatureFlags.java:678]: How would this work if the extension supports multiple (connected) CCloud connections?

File: src/main/java/io/confluent/idesidecar/restapi/featureflags/HttpFlagEvaluationProvider.java
Comment (flippingbits) [inline src/main/java/io/confluent/idesidecar/restapi/featureflags/HttpFlagEvaluationProvider.java:28]: Can you `@Inject` the `WebClientFactory` object? Otherwise, we'll end up using a different instance than the rest of the application, which will be problematic if, for instance, the user uses custom SSL settings.

File: src/main/java/io/confluent/idesidecar/restapi/util/OperatingSystemType.java
Comment (flippingbits) [inline src/main/java/io/confluent/idesidecar/restapi/util/OperatingSystemType.java:16]: Can you think of a situation in which we need to determine the OS of a system other than the _current_ one?

**Review by flippingbits — COMMENTED** [review]
Great work, @rhauch! I have a few questions/comments. Otherwise, your PR looks good to me.

File: src/main/java/io/confluent/idesidecar/restapi/exceptions/ExceptionMappers.java
Comment (rhauch) [reply src/main/java/io/confluent/idesidecar/restapi/exceptions/ExceptionMappers.java:104]: This only happens if the sidecar is unable to parse the LD response, so I would consider it an internal error. But as you suggested, I will add 500 to the OpenAPI spec.

File: src/main/java/io/confluent/idesidecar/restapi/featureflags/FeatureFlags.java
Comment (rhauch) [reply src/main/java/io/confluent/idesidecar/restapi/featureflags/FeatureFlags.java:347]: Good question. The main reason is that I originally intended `FlagId` to have literals for all of the feature flags **used in the sidecar**, and wanted to keep the REST API flexible by accepting any string ID. This way the VS Code extension could use flags that are not included in the `FlagId` enumeration.

I guess I still think this is the more flexible route, but I'm open to feedback. Regardless, I will update the JavaDoc on this method to better describe the difference between this method and the others that take a `FlagId` parameter rather than `String`.

File: src/main/java/io/confluent/idesidecar/restapi/featureflags/FeatureFlags.java
Comment (rhauch) [reply src/main/java/io/confluent/idesidecar/restapi/featureflags/FeatureFlags.java:678]: Yeah, that's a big challenge that I chose to not address here, because I think there's no good solution as the feature flags almost certainly will not be evaluated within the context of a specific connection ID. 
IOW, as implemented, the feature flag evaluations will always reflect the most-recently connected CCloud connection.

I guess I justified this behavior with the current extension UX that has at most one CCloud connection, and that we'll probably stick with that for the foreseeable future.

File: src/main/java/io/confluent/idesidecar/restapi/featureflags/HttpFlagEvaluationProvider.java
Comment (rhauch) [reply src/main/java/io/confluent/idesidecar/restapi/featureflags/HttpFlagEvaluationProvider.java:28]: Good question. The `HttpFlagEvaluationProvider` is not a bean, and I'd rather keep it not a bean. However, I take your point, and will change this so that the `WebClientFactory` instance is injected in the lowest-level bean (`FeatureFlags`) and passed down into the other objects.

File: src/main/java/io/confluent/idesidecar/restapi/util/OperatingSystemType.java
Comment (rhauch) [reply src/main/java/io/confluent/idesidecar/restapi/util/OperatingSystemType.java:16]: This `from(...)` method doesn't assume the given properties are system properties; I just called the variable that. So technically this `from(...)` method does not assume it's the current system.

The `SidecarInfo` bean's default constructor does use the `System.getProperty(...)` method to determine the OS type, but for testing purposes there is another constructor that the tests can pass other non-`System` properties and verify the behavior of `SidecarInfo` and `OperatingSystemType`.

File: src/main/java/io/confluent/idesidecar/restapi/exceptions/ParsingFeatureFlagsFailedException.java
Comment (rhauch) [reply src/main/java/io/confluent/idesidecar/restapi/exceptions/ParsingFeatureFlagsFailedException.java:12]: Good point. I'm going to refactor this, by moving the record into the `HttpFeatureFlagProvider` and renaming it (since it represents the error response from LD and shouldn't exist outside of that provider, and because that's where the `FeatureFlagFailureException` are instantiated). I'll also change the `FeatureFlagFailureException` to have a `code` field rather than the record type.

File: src/main/java/io/confluent/idesidecar/restapi/featureflags/HttpFlagEvaluationProvider.java
Comment (rhauch) [reply src/main/java/io/confluent/idesidecar/restapi/featureflags/HttpFlagEvaluationProvider.java]: Made this change.

File: src/main/java/io/confluent/idesidecar/restapi/exceptions/ParsingFeatureFlagsFailedException.java
Comment (rhauch) [reply src/main/java/io/confluent/idesidecar/restapi/exceptions/ParsingFeatureFlagsFailedException.java]: Made this change.

File: src/main/java/io/confluent/idesidecar/restapi/exceptions/ExceptionMappers.java
Comment (rhauch) [reply src/main/java/io/confluent/idesidecar/restapi/exceptions/ExceptionMappers.java:104]: Made this change.

File: src/main/java/io/confluent/idesidecar/restapi/util/OperatingSystemType.java
Comment (flippingbits) [reply src/main/java/io/confluent/idesidecar/restapi/util/OperatingSystemType.java:16]: I was asking because I couldn't find an usage of the `from(...)` method with `Properties` other than the System ones so I thought only having `OperationSystemType::current` is enough. I guess it's fine to have the `from(...)` method if we need it for tests later on.

File: src/main/java/io/confluent/idesidecar/restapi/featureflags/FeatureFlags.java
Comment (flippingbits) [reply src/main/java/io/confluent/idesidecar/restapi/featureflags/FeatureFlags.java:692]: Makes sense! I also don't think that adding support for multiple CCloud connections justifies the effort here. I just want us to keep in mind that adding support for multiple CCloud connections will require a bigger effort on the sidecar side after merging this PR.

File: src/main/java/io/confluent/idesidecar/restapi/featureflags/FeatureFlags.java
Comment (flippingbits) [reply src/main/java/io/confluent/idesidecar/restapi/featureflags/FeatureFlags.java:361]: Thanks for the clarification!

File: src/main/java/io/confluent/idesidecar/restapi/exceptions/ExceptionMappers.java
Comment (flippingbits) [reply src/main/java/io/confluent/idesidecar/restapi/exceptions/ExceptionMappers.java:104]: Using `Status.INTERNAL_SERVER_ERROR` makes a lot of sense here. Can you please also update the `Response.status` in the exception mapper function?

**Review by flippingbits — COMMENTED** [review]
Thanks for the update, @rhauch! I added a few more minor comments/questions.

File: src/main/java/io/confluent/idesidecar/restapi/exceptions/ExceptionMappers.java
Comment (rhauch) [reply src/main/java/io/confluent/idesidecar/restapi/exceptions/ExceptionMappers.java:104]: Actually, the `FeatureFlagFailureException` are logged, since that means we had trouble parsing the LD response, but that parsing and this error only occur on background threads and does not occur on REST API request threads. So I'm going to remove this exception mapping.

File: src/main/java/io/confluent/idesidecar/restapi/util/OperatingSystemType.java
Comment (rhauch) [reply src/main/java/io/confluent/idesidecar/restapi/util/OperatingSystemType.java:16]: Not directly. Most of the test cases in `SidecarInfoTest` do **not** use `System::getProperty(...)` or `System::getenv(...)`, and instead use a map with `Map::get(...)`. See [here](https://github.com/confluentinc/ide-sidecar/pull/76/files#diff-d0cceb0329defee9693d4c44a3c86bd965b6b11774b3e8657a17b56fd140423dR25-R50) and [here](https://github.com/confluentinc/ide-sidecar/pull/76/files#diff-d0cceb0329defee9693d4c44a3c86bd965b6b11774b3e8657a17b56fd140423dR52-R55).

File: src/main/java/io/confluent/idesidecar/restapi/application/SidecarInfo.java
Comment (rhauch) [reply src/main/java/io/confluent/idesidecar/restapi/application/SidecarInfo.java:39]: Yeah, I'm not sure it really matters, but I thought it was beneficial to specify my desired semantics.

File: src/main/java/io/confluent/idesidecar/restapi/featureflags/FeatureFlags.java
Comment (rhauch) [reply src/main/java/io/confluent/idesidecar/restapi/featureflags/FeatureFlags.java:144]: Is there a good reason to use `@Singleton` instead? Using `@ApplicationScoped` and `@Startup` seems to work, and `@Startup` is required in order for the flags to be evaluated upon startup.

Comment (flippingbits) [reply src/main/java/io/confluent/idesidecar/restapi/featureflags/FeatureFlags.java:144]: I think it's fine to keep it if it works. Using `@ApplicationScoped` and `@Startup` at the same time just felt a bit contradictory.

**Comment by rhauch** [comment]
Thanks for the review and feedback, @flippingbits. I've made a few changes based upon your suggestions. Would you please take another look?

**Review by flippingbits — APPROVED** [review]
Excellent work, @rhauch!

**Comment by rhauch** [comment]
Thanks, @flippingbits. I've added the `on-hold` label since I'm waiting for a few more reviews before I merge this.

**Comment by rhauch** [comment]
Actually, still on-hold because we're waiting for some final corrections to the `IDE` project in Confluent's LaunchDarkly.

**Pull request merged**
